types/src/proof/ensure_proof.rs

```diff
@@ -0,0 +1,31 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+//! This module declares the macro ensure_proof!,
+//! which checks a condition for a proof and throws a ProofError
+//! if the proof fails.
+//! Currently, this macro only supports SparseMerkleProofs
+
+use crate::proof::{definition::Proof, proof_error::ProofError};
+use failure::Error;
+
+macro_rules! ensure_proof {
+    ($proof:ident, $cond:expr, $e:expr) => {
+        if Proof::is_proof($proof) {
+            if !($cond) {
+                return Err(Error::from(ProofError::new($proof.to_string(), $e.to_string())));
+            }
+        } else {
+            ensure!($cond, $e);
```

Comment:
This line is confusing... I think the macro `ensure_proof` should not be acted like `ensure!` when seeing something that is not a proof. And the first param looks weird to me...Why don't you define it as a function but a macro? IMO there is not much boilerplate code that necessitates a macro.